lib/src/tools/format_tool.dart

```diff
@@ -0,0 +1,418 @@
+  /// known dart project directories (`benchmark/`, `bin/`, `example/`, `test/`,
+  /// `tool/`, `web/`) as well as dart files in the root.
+  List<Glob> include;
+
+  @override
+  FutureOr<int> run([DevToolExecutionContext context]) {
+    context ??= DevToolExecutionContext();
+    final execution = buildExecution(context,
+        configuredFormatterArgs: formatterArgs,
+        defaultMode: defaultMode,
+        exclude: exclude,
+        formatter: formatter,
+        include: include);
+    return execution.exitCode ?? runProcessAndEnsureExit(execution.process);
+  }
+
+  @override
+  Command<int> toCommand(String name) => DevToolCommand(name, this,
+      argParser: ArgParser()
+        ..addSeparator('======== Formatter Mode')
+        ..addFlag('overwrite',
+            abbr: 'w',
+            negatable: false,
+            help: 'Overwrite input files with formatted output.')
+        ..addFlag('dry-run',
+            abbr: 'n',
+            negatable: false,
+            help: 'Show which files would be modified but make no changes.')
+        ..addFlag('assert',
+            abbr: 'a',
+            negatable: false,
+            help:
+                'Assert that no changes need to be made by setting the exit code '
+                'accordingly.\nImplies "--dry-run" and "--set-exit-if-changed".')
+        ..addSeparator('======== Other Options')
+        ..addOption('formatter-args',
+            help: 'Args to pass to the "dartfmt" process.\n'
+                'Run "dartfmt -h -v" to see all available options.'));
+}
+
+/// A declarative representation of an execution of the [FormatTool].
+///
+/// This class allows the [FormatTool] to break its execution up into two steps:
+/// 1. Validation of confg/inputs and creation of this class.
+/// 2. Execution of expensive or hard-to-test logic based on step 1.
+///
+/// As a result, nearly all of the logic in [FormatTool] can be tested via the
+/// output of step 1 (an instance of this class) with very simple unit tests.
+class FormatExecution {
+  FormatExecution.exitEarly(this.exitCode) : process = null;
+  FormatExecution.process(this.process) : exitCode = null;
+
+  /// If non-null, the execution is already complete and the [FormatTool] should
+  /// exit with this code.
+  ///
+  /// If null, there is more work to do.
+  final int exitCode;
+
+  /// A declarative representation of the formatter process that should be run.
+  ///
+  /// This process' result should become the final result of the [FormatTool].
+  final ProcessDeclaration process;
+}
+
+/// Modes supported by the dart formatter.
+enum FormatMode {
+  // dartfmt -n --set-exit-if-changed
+  assertNoChanges,
+  // dartfmt -n
+  dryRun,
+  // dartfmt -w
+  overwrite,
+}
+
+/// Available dart formatters.
+enum Formatter {
+  // The formatter provided via the Dart SDK.
+  dartfmt,
+  // The formatter provided via the `dart_style` package.
+  dartStyle,
+}
```

Comment:
Is it worth making both of these an option? Is `dartfmt` not an alias to the other?

Reply:
`dartfmt` does run `dart_style:format`, but with the version of dart_style that is packaged with the Dart SDK. Some packages depend on `dart_style` directly and run that exe, which allows them to pick a specific version or pull in newer versions that are available sooner than they would be via an SDK release.

Reply:
Another way to look at the difference – `dartfmt` works in any dart project, but `pub run dart_style:format` requires `dart_style` be a direct dependency of the current project.